### PR TITLE
Package hierarchy transformer

### LINKED ISSUE #53: Fix package hierarchy generation
Currently packages are not nested in the “list” on the LHS.

The list is also folded on object pages.

### PR DESCRIPTION
**Pull request by pikinier20**
fixes #53

### CODE REVIEW

File: src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            other
+        }
+    }
+
+    def rearangePackagePages(page: ModulePageNode): ModulePageNode = {
+        val packagePages = page.getChildren.asScala.collect{ case p: PackagePageNode => p }.toList
+
+
+        def isParent(possibleParent: Seq[String], comparedChildren: Seq[String]): Boolean = {
+            possibleParent.zip(comparedChildren).forall(_ == _)
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala]: why not `possibleParent == comparedChildren`?

**Review by romanowski — CHANGES_REQUESTED** [review]
I think the nested packages should use single not full name, so we should have `sclala` and `runtime` not `scala` and `scala.runtime` on the pic below
Screenshot from 2020-08-13 15-35-07

File: src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala (hunk 1 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala]: possibleParent and comparedChildren are of different sizes, so there would be no equality. Zip ignores remaining elements of longer list, so then I can compare elem by elem

**Comment by pikinier20** [comment]
Made names of package pages relative

**Comment by abgruszecki** [comment]
@romanowski I think we can merge this, right? It'd be nice to have this before the Contributors post.

File: src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala (hunk 1 above)
Comment (romanowski) [reply src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala]: you can use `startsWith` in such case

File: src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                childPage.getChildren
+            )
+
+        def buildPackageTree(
+            depth: Int, 
+            remaining: Seq[(Seq[String], PackagePageNode)], 
+            processsed: Seq[(Seq[String], PackagePageNode)]
+        ): Seq[PackagePageNode] = {
+            val (currentDepth,rest) = remaining.partition((tokens, page) => tokens.size == depth)
+            val newProcessed = currentDepth.map( (tokens, page) =>
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala]: Nitpick (one below is much more readable):
```
val newProcessed = currentDepth.map{  (tokens, page) => 
                val newPage = page.modified(
                        page.getName,
                        (processsed
                            .filter((childTokens, child) => isParent(tokens, childTokens))
                            .map((childTokens, child) => relativePageName(tokens, childTokens, child))
                        ++ page.getChildren.asScala).asJava,
                    )
                (tokens, newPage)
            }
```

Comment (pikinier20) [reply src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala]: Changed

File: src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala (hunk 1 above)
Comment (pikinier20) [reply src/main/scala/dotty/dokka/PackageHierarchyTransformer.scala]: That's a good point, changed.

**Review by romanowski — APPROVED** [review]

**Pull request merged**